src/dswx_sar/pre_processing.py

```diff
@@ -0,0 +1,407 @@
+        else:
+            self.ycoord_rtc, self.xcoord_rtc = self.read_x_y_array_geotiff(rtc_file_name)
+            reftif = gdal.Open(rtc_file_name)
+            proj = osr.SpatialReference(wkt=reftif.GetProjection())
+            self.epsg = proj.GetAttrValue('AUTHORITY',1)
+            reftif = None
+        self.scratch_dir = scratch_dir
+        print('ancillary_reloation', self.epsg)
+
+    def _relocate(self,
+                  ancillary_file_name,
+                  relocated_file_str,
+                  method='near'):
+
+        """ resample image
+
+        Parameters
+        ----------
+        ancillary_file_name : str
+            file name of ancilary data
+        relocated_file_str : str
+            file name of output
+        method : str
+            interpolation method
+
+        Returns
+        -------
+        resampled_slc or slc_demodulate: numpy.ndarray
+            numpy array of bandpassed slc
+            if resampling is True, return resampled slc with bandpass and demodulation
+            if resampling is False, return slc with bandpass and demodulation without resampling
+        meta : dict
+            dict containing meta data of bandpassed slc
+            center_frequency, rg_bandwidth, range_spacing, slant_range
+        """
+
+        self.interpolate_gdal(str(self.rtc_file_name),
+                         ancillary_file_name,
+                         os.path.join(self.scratch_dir, relocated_file_str),
+                         method)
+
+    def interpolate_gdal(self, ref_file, input_tif_str, output_tif_str, method, epsg=None):
+        print(f"> gdalwarp {input_tif_str} -> {output_tif_str}")
+
+        # get reference coordinate and projection
+        if  h5py.is_hdf5(ref_file):
+            with h5py.File(ref_file, 'r') as f4:
+                try:
+                    im = np.array(f4['/science/LSAR/GCOV/grids/frequencyA/VVVV'])
+                except:
+                    im = np.array(f4['/science/LSAR/GCOV/grids/frequencyA/HHHH'])
+                ysize, xsize = np.shape(im)
+                lon0 = np.array(f4['/science/LSAR/GCOV/grids/frequencyA/xCoordinates'])
+                lat0 = np.array(f4['/science/LSAR/GCOV/grids/frequencyA/yCoordinates'])
+                epsg_output = np.array(f4['/science/LSAR/GCOV/grids/frequencyA/projection'])
+        else:
+            reftif = gdal.Open(ref_file)
+            lat0, lon0 = self.read_x_y_array_geotiff(ref_file)
+            xsize = reftif.RasterXSize
+            ysize = reftif.RasterYSize
+            geotransform = reftif.GetGeoTransform()
+            proj = osr.SpatialReference(wkt=reftif.GetProjection())
+            epsg_output = proj.GetAttrValue('AUTHORITY',1)
+            xspacing = geotransform[1]
+            yspacing = geotransform[5]
+
+            reftif = None
+            del reftif
+
+        if ( len(lat0) != ysize ) and ( len(lon0) != xsize ):
```

Comment:
What is going to happen when `len(lat0) != ysize` is `True` but `len(lon0) != xsize` is `False`?

Reply:
This was observed when the RTC was stored in HDF5 file. We may need to review this part when we are going to deal with the NISAR RTC products.